### PR TITLE
[WIP] Cache profiling

### PR DESCRIPTION
**Pull request by jcrist**
Tracks the contents of the cache, and plots it against time. Default metric is just to count by task type, but other metrics can be used (for example `chest.core.nbytes`).

screen shot 2015-10-14 at 1 57 42 pm

TODO:
- [ ] Needs tests
- [ ] Still not happy with the plotting design

### CODE REVIEW

**Comment by jcrist** [comment]
I'm still not happy with the plot, but I've mucked around in bokeh far too much today. What I originally wanted was a stacked area plot, so the peak of the top area would show the total count. However, bokeh no longer has area plots (deprecated old charts), so I had to roll my own. Then I realized that the lines really should be stepped, because the changes in cache size happen at discrete points. Could definitely get that done, but I need a break for now...

If anyone (@mrocklin?) has thoughts on what the plot should look like (or if this information is useful for you) let me know.

**Comment by mrocklin** [comment]
Oh cool, so the line colors match up with the tasks above?

**Comment by jcrist** [comment]
Yeah. I'd like them to be areas though. This would also help with the lighter colors (one of the lines is hardly visible), and allow hovering to be more useful.

File: dask/diagnostics/profile.py [hunk 1]
```diff
@@ -210,3 +210,110 @@ def run(self):
...
+    22
+
+    >>> prof.results    # doctest: +SKIP
+    [CacheData('y', (add, 'x', 10), 1, 1435352238.48039, 1435352238.480655),
+     CacheData('z', (mul, 'y', 2), 1, 1435352238.480657, 1435352238.480803)]
+
+    The default is to count each task (``metric`` is 1 for all tasks). Other
+    functions may used as a metric instead through the ``metric`` keyword. For
+    example, the ``nbytes`` function found in ``chest`` can be used to measure
+    the number of bytes in the cache.
```
Comment (mrocklin) [inline dask/diagnostics/profile.py]: FYI the nbytes function in cachey is a bit more accurate (I think)

**Comment by mrocklin** [comment]
OK, I had a chance to play with this on a real problem.  Feedback:

1.  The relative quantity of different kinds of outputs can vary dramatically.  The ability to zoom on the y-axis would be very useful.
2.  The mouse-over / tool-tip is really handy.  Ideally it'd be easier to hit.  I think I'm aiming for just a pixel or two (I'm not sure how easy this is to fix without serious bokeh magic though.)
3.  If I use my own metric what happens to the y-label?  Does it remain count?  If so then perhaps we should remove this and leave it ambiguous.

**Comment by jcrist** [comment]
Fixed 1. Number 2 is tricky, I've asked some bokeh people for help here - hovering on lines doesn't work, and it seems tricky to make the hover only snap to one point. For 3, the y-label is either the `metric_name` keyword argument, or `metric.__name__`, in that order. Default is `count`, for the default function.

I'm tempted to merge this, and fix the hover later. I suspect a few changes will have to be made to this plot before we're happy with it.

**Comment by mrocklin** [comment]
Punting on hover makes sense to me.  After using it more extensively I'm pretty content with 3

**Comment by jcrist** [comment]
Cool. Failures are due to a periodic test failure. Merging.

**Pull request merged**